### ISSUE TITLE
ML Experiment Tracking and Management

### ISSUE THREAD

**Issue by dyt811** [ivadomed/ivadomed#1010]
### Discussed in https://github.com/ivadomed/ivadomed/discussions/1004

<div type='discussions-op-text'>

<sup>Originally posted by **naga-karthik** December 13, 2021</sup>
The aim of this post is to open a stage for discussing various alternatives for ML experiment tracking. 

### Motivation:

As we train more and more DL models, keeping track of the models (i.e. how many filters were used, what learning rate, epochs, etc.) becomes increasingly difficult. The brute-force way is to manually add entries onto a spreadsheet but there are several much better alternatives that make this task a lot easier.

### Trackers Explored:

- [Weights and Biases (wandb.ai)](https://wandb.ai/site): This was tested (by @naga-karthik) while working on a course project. My overall experience has been good. There are several useful features, for e.g. grouping experiments into "Runs", visualization of all loss curves, logging of hyperparameters and gradients. Most importantly, everything can be found online (on the user's wandb profile), which avoids running it on localhost (as it is done on Tensorboard). 
   - While it is easy to use on a personal basis (an account can be created free of charge), I am unaware of the potential issues it might create when we want to integrate it into the ivadomed system. Particularly, will it be considered as an "enterprise" account, or, are the users who will use ivadomed expected to create wandb accounts as a prerequisite? Need some discussion here.
   - [Here](https://wandb.ai/site/pricing) are the pricing details for wandb 
  - Pros:
    - @uzaymacar and @naga-karthik are familiar with it (maybe others in the lab?) and can fill up the pros/cons

- [Comet.ml](https://www.comet.ml/site/): has been tried @dyt811 
   -  Comet.ml has apparently good integration with a wide variety of ML run time environments/framewokrs: troch/tensorflow/SKLearn/MLFlow, etc. 
   - Key commands for logging [Comet.ml cheat sheet: supercharge your machine learning experiment management](https://medium.com/comet-ml/comet-ml-cheat-sheet-supercharge-your-machine-learning-experiment-management-7786e08f8e9e) 
   - [Hyperparameters search](https://www.comet.ml/docs/python-sdk/introduction-optimizer/)
      - Also check out Optuna also for optimization
   - Pro: 
      - Off the shelf, centralized web portal to get started with API keys within literally minutes to log metrics/data
      - Good for metrics/parameters/experiments tracking with minimum amount of code. 
      - compatible with multiple experimental tracking solutions
      - has built in optimizer
      - Probably more commercial support if you are on paid tier? 
      - Free for academia affiliated researchers
   - Con: 
      - Expensive if you have to buy it (a few hundred per month per seat), 
      - not open source
      - not locally deployable AFAIK. 

- [MLFlow](https://mlflow.org): has been around for a while too
    - Pro: 
       - Free. Open source
       - Can be used to reproduce/replicate simple experiments (have not tried run on more complicated deep learning)
       - Probably more community support? 
    - Con:
       - Lack an off the shelf centralized SAAS solution. You will have to host your own (e.g. kind of like a tensorboard portal)
       - Slightly less feature than the Comet.ml solution in the track space. Not 100% sure. ? 

### Future:

In my opinion, having an experiment tracker (that can be accessed online without localhost) is definitely a plus point. It turns out [this issue](https://github.com/ivadomed/ivadomed/issues/801) already brought up this topic and some more information on this can be found here. I hope that this post will motivate more discussions and possibly others can chip in with other (newer) experiment trackers that they might have used.

</div>

**Comment by uzaymacar** [comment]
There is also [SigOpt](https://sigopt.com) who has a partnership with Mila and apparently OpenAI also uses it. Meanwhile, Facebook seems to have adopted wandb as inferred from [here](https://github.com/facebookresearch/vissl/issues/268).

**Comment by naga-karthik** [comment]
Here's another idea: SigOpt is more of a tool for Bayesian hyperparameter optimization as is discussed here: [blog 1](https://sigopt.com/blog/using-sigopt-to-optimize-neural-networks-in-biology/) and [blog 2](https://sigopt.com/blog/academic-interview-with-brady-neal-from-mila/) (See the section "Why did you select SigOpt to optimize these models?"). It seems particularly useful when you have a large no. of hyperparams (genomics/biology examples come to mind) to tune and the best way to do it at scale is by using a Bayesian approach (which is where SigOpt excels at). Here's another open-source package [DeepDIVA](https://diva-dia.github.io/DeepDIVAweb/articles/run-sigopt/) (like ivadomed) which incorporated SigOpt inside of it. Briefly, they use TensorBoard + SigOpt combination for experiment tracking and hyperparameter optimization.

Choosing between wandb and SigOpt: The answer to this depends on what we want for the users. If we want to let them optimize hyperparams themselves then incorporating only wandb into ivadomed should be the way. Or else, we can also have a wandb + sigopt combination such that we utilize the best for both tasks (i.e. tracking and hyperparam optimization) but this surely means more maintenance from our side.

**Comment by naga-karthik** [comment]
From what I am seeing, it seems that wandb also has hyperparameter optimization, making it a compelling choice! Here's another list of Pros for wandb:

1. Easily integrable with Pytorch and Pytorch Lightning
2. Hyperparameter optimizing ("Sweeps") (Note: not tried it myself)
3. Dataset and model versioning
4. Free for personal use (means that the users just need to create a wandb profile and get an api-token to start tracking) 

Unrelated: wandb also recently closed its Series C funding of >100M$ ([source](https://finance.yahoo.com/news/weights-biases-raises-series-c-130000883.html))

**Comment by kanishk16** [comment]
Since I am familiar with wandb, I would like to add a few more Pros for wandb:

1. Tracks System Usage (GPU - Power Usage, Temp, Utilization, etc.)
2. Extensive Tutorials available (as a user, reduces your learning curve)

As far as pricing is concerned, there is an option available for academic research teams which can be explored.

**Comment by jcohenadad** [comment]
heads up: @naga-karthik is working on an implementation-- and i also see that @dyt811 is assigned to this issue-- we should absolutely try avoiding several ppl working in parallel without communicating, so please @naga-karthik can you open a PR (in draft mode) so the team can start chipping in, give feedback, brainstorm, etc.

**Comment by naga-karthik** [comment]
Yes! Thank you for the suggestion @jcohenadad! I have created a draft PR #1069 that aims to resolve this. Looking forward to your suggestions! Sorry if anyone else was working on this already.
